Move crm_simulate into libpacemaker

clumens opened the pull request:
This is just the beginning of this.  I think there's a lot more to do.  I've got several questions:

* Should I be more aggressive with moving things into libpacemaker?  There are whole blocks in the main function that look like they could get moved out to allow other tools to do what crm_simulate does.  At the same time, there's a lot of input/output managing mixed up with the real code.
* Can modify_configuration and run_simulation be renamed to follow along with naming schemes used elsewhere?
* setup_input really looks like code that probably exists elsewhere, or at least should exist elsewhere and be shared.  Does this look familiar to anyone?
* Do the pcmk__ functions in lib/pacemaker/pcmk_simulate.c need a public equivalent eventually?  If so, we'll have to think of a way to handle the output object like the other public libpacemaker functions.  It's a little more complicated here because we are relying on data_set to have an output object.  I was thinking the public functions could create the output object, set it on data_set, and then call the private function.  This isn't an exact match with the other public functions, but it's pretty close.

kgaillot commented:
Looks great, thanks.

> This is just the beginning of this. I think there's a lot more to do. I've got several questions:
> 
>     * Should I be more aggressive with moving things into libpacemaker?  There are whole blocks in the main function that look like they could get moved out to allow other tools to do what crm_simulate does.  At the same time, there's a lot of input/output managing mixed up with the real code.

Ideally the tools will just have argument processing, but sometimes that's not practical.

My gut feeling is that crm_simulate's -L/-x/-p options should be deprecated in favor of using CIB_file, so I'd rather not spread their handling (mainly setup_input()) anywhere else. I'm not ready to go as far as deprecating them, though.

Most of main() after the argument parsing minus setup_input() could be one big high-level libpacemaker function (well one private and one public). How to pass a zillion arguments would be the main question -- should we create simulation flags to replace the booleans, and a struct or two for most of the rest? 
 
>     * Can modify_configuration and run_simulation be renamed to follow along with naming schemes used elsewhere?

Yes, they're internal.

I haven't had a chance yet to go through any libraries other than libcrmcommon to apply current guidelines. libpacemaker is almost entirely internal so it's a great candidate but a ton of work.

>     * setup_input really looks like code that probably exists elsewhere, or at least should exist elsewhere and be shared.  Does this look familiar to anyone?

setup_input() is a little strange, actually. It reads the CIB (from whatever source), then writes it to a file and sets CIB_file to that. I'm guessing that's to respect whether --in-place is used or not. (BTW I don't see anything limiting --in-place to when -x is used, I'm guessing it would choke if used with -p or -L, or maybe try to overwrite the live CIB with -L ...)

I'd leave most of setup_input() in crm_simulate.c, just because I feel like the input handling could benefit from an overhaul someday. However the cib new-signon-query sequence is used in a few places, so that could be a convenience function in libcib (probably internal for now).

>     * Do the pcmk__ functions in lib/pacemaker/pcmk_simulate.c need a public equivalent eventually?  If so, we'll have to think of a way to handle the output object like the other public libpacemaker functions.  It's a little more complicated here because we are relying on data_set to have an output object.  I was thinking the public functions could create the output object, set it on data_set, and then call the private function.  This isn't an exact match with the other public functions, but it's pretty close.

I don't think they need individual public equivalents, but one high-level public function as described above would be appropriate.

kgaillot commented:
@tomjelinek , would the proposed crm_simulate output changes here cause problems for pcs? See examples in files changed

clumens commented:
> I'd leave most of setup_input() in crm_simulate.c, just because I feel like the input handling could benefit from an overhaul someday. However the cib new-signon-query sequence is used in a few places, so that could be a convenience function in libcib (probably internal for now).

There ended up being about a half dozen places where this change could be made, which wasn't as many as I was hoping for but oh well.  I've added a patch to make this into a function.

tomjelinek commented:
> @tomjelinek , would the proposed crm_simulate output changes here cause problems for pcs? See examples in files changed

If I understand it correctly, this will break [`pcs.resource.resource_relocate_show`](https://github.com/ClusterLabs/pcs/blob/15ca9f88520b6d919a14ef6121c69643d2c244ce/pcs/resource.py#L3290) function which parses the plaintext output of crm_simulate. It doesn't seem to be difficult to update the function so that it's able to process both the old and the new output, though.

kgaillot commented:
> > @tomjelinek , would the proposed crm_simulate output changes here cause problems for pcs? See examples in files changed
> 
> If I understand it correctly, this will break [`pcs.resource.resource_relocate_show`](https://github.com/ClusterLabs/pcs/blob/15ca9f88520b6d919a14ef6121c69643d2c244ce/pcs/resource.py#L3290) function which parses the plaintext output of crm_simulate. It doesn't seem to be difficult to update the function so that it's able to process both the old and the new output, though.

As of 2.1.0, crm_simulate supports --output-as=xml, so it would be better to switch to that when available rather than update the expected text output.

I'm wary of merging this without pcs having support ready. I hate to do it, but maybe we should add another flag to the cluster-status message for whether to show this info crm_mon-style or crm_simulate-style. We could drop it after pcs can handle XML.

clumens commented:
Updated with a bunch more patches, but I'm still marking this as a draft for now.  I need to do testing beyond just what's in CTS, at least.

kgaillot reviewed (commented):
This looks great, really helps readability. I think we should figure out how to keep the output the same so as not to break pcs though, hopefully being able to drop that difference eventually.

We should have a public equivalent of pcmk__simulate(), which means making the events struct and sim flags enum public too. The long-term goal is to wrap a Python API around the high-level APIs so projects like pcs can call them natively rather than invoke the command-line tools. Other bindings could be added as needed.

clumens commented:
> I'm wary of merging this without pcs having support ready. I hate to do it, but maybe we should add another flag to the cluster-status message for whether to show this info crm_mon-style or crm_simulate-style. We could drop it after pcs can handle XML.

Probably the easiest thing to do would just be to drop "Use the cluster-status message in crm_simulate" and then make whatever other little changes are required in later patches to adapt them.  We could add a TODO comment as well, for when it's safe to combine the code.

clumens commented on include/pcmki/pcmki_simulate.h:
```diff
@@ -13,6 +13,17 @@
+    sim_show_utilization = 1 << 4,
+    sim_simulate         = 1 << 5,
+    sim_store            = 1 << 6,
+} pcmk__sim_flags;
```
I thought about doing that and then didn't.  I should have just gone for it.  Because we're going to want the flags to be public, I assume there should be only a single underscore on all those?  I'd additionally want to define them in include/pacemaker.h instead, right?  I think all the include/pcmki/ headers are supposed to be internal.

kgaillot replied:
Correct on all counts

clumens replied:
Fixed.

clumens commented:
> We should have a public equivalent of pcmk__simulate(), which means making the events struct and sim flags enum public too. 

I'm not fully sold on the pcmk__sim_events_t name.  I guess we do use "event" in various other places, so maybe it's fine.  I was thinking about calling it pcmk__sim_actions_t instead.

> The long-term goal is to wrap a Python API around the high-level APIs so projects like pcs can call them natively rather than invoke the command-line tools. Other bindings could be added as needed.

That does seem like a very long term goal.

clumens commented on tools/crm_simulate.c:
```diff
@@ -113,48 +103,48 @@ live_check_cb(const gchar *option_name, const gchar *optarg, gpointer data, GErr
 node_down_cb(const gchar *option_name, const gchar *optarg, gpointer data, GError **error) {
     options.modified++;
-    options.node_down = g_list_append(options.node_down, (gchar *) g_strdup(optarg));
+    options.events.node_down = g_list_append(options.events.node_down, (gchar *) g_strdup(optarg));
```
Hah, that's been there since I converted crm_simulate to glib over a year ago.  Fixed.

clumens commented on tools/crm_simulate.c:
```diff
@@ -773,20 +760,20 @@ main(int argc, char **argv)
+    g_list_free_full(options.events.ticket_activate, g_free);
     free(options.use_date);
-    free(options.watchdog);
+    free(options.events.watchdog);
```
Added a patch to the end of the series.

kgaillot commented:
> > I'm wary of merging this without pcs having support ready. I hate to do it, but maybe we should add another flag to the cluster-status message for whether to show this info crm_mon-style or crm_simulate-style. We could drop it after pcs can handle XML.
> 
> Probably the easiest thing to do would just be to drop "Use the cluster-status message in crm_simulate" and then make whatever other little changes are required in later patches to adapt them. We could add a TODO comment as well, for when it's safe to combine the code.

Ah I did like that change though :) I guess it's for the best ...

kgaillot commented:
> I'm not fully sold on the pcmk__sim_events_t name. I guess we do use "event" in various other places, so maybe it's fine. I was thinking about calling it pcmk__sim_actions_t instead.

Action tends to have a more specific meaning in pacemaker. Event is actually closer. Or maybe pcmk_simulations_t or pcmk_injections_t?

clumens commented:
> > Probably the easiest thing to do would just be to drop "Use the cluster-status message in crm_simulate" and then make whatever other little changes are required in later patches to adapt them. We could add a TODO comment as well, for when it's safe to combine the code.
> 
> Ah I did like that change though :) I guess it's for the best ...

Actually, I've come up with a way that gets us down to a single digit number of test cases changed.  I'll post a new patch in the next day or two, and we can run it by the pcs guys and see if it's good enough.  If we want to have zero output changes, maybe I can come up with some other ways to do that.

clumens commented on lib/cib/cib_utils.c:
```diff
@@ -785,3 +785,41 @@ cib_apply_patch_event(xmlNode *event, xmlNode *input, xmlNode **output,
+    } else {
+        *cib = cib_new();
+        cib_conn = *cib;
+    }
```
Is there anything preventing us from modifying cib_remote_new, cib_native_new, cib_new_variant, etc. and checking for/returning NULL right now?

kgaillot replied:
Nothing but time

clumens commented on tools/crm_verify.c:
```diff
@@ -148,28 +148,17 @@ main(int argc, char **argv)
+        } else if (rc != pcmk_rc_ok) {
+            g_set_error(&error, PCMK__RC_ERROR, rc, "Live CIB query failed: %s", pcmk_rc_str(rc));
+            goto done;
         }
```
Related to error handling, note the commit message for 2ef68d5185440b8509ea9f6ecbc854fc526bec0e and what has happened to all the GError usage there.

kgaillot replied:
I do think the functions should return standard error codes, and the mains should set GError from it if desired. But if we're setting GError differently for different codes, we're probably using the wrong codes -- pcmk_rc_str() should give something appropriate.

clumens replied:
The problem seems to be more that the simulate function could return (for instance) some errno value that could either be because writing the dot file failed, or because writing the XML file failed.  Without the GError, the caller won't know which the errno corresponds to.

kgaillot replied:
Right, that means the simulate function isn't returning sufficient codes

kgaillot commented:
> > > Probably the easiest thing to do would just be to drop "Use the cluster-status message in crm_simulate" and then make whatever other little changes are required in later patches to adapt them. We could add a TODO comment as well, for when it's safe to combine the code.
> > 
> > 
> > Ah I did like that change though :) I guess it's for the best ...
> 
> Actually, I've come up with a way that gets us down to a single digit number of test cases changed. I'll post a new patch in the next day or two, and we can run it by the pcs guys and see if it's good enough. If we want to have zero output changes, maybe I can come up with some other ways to do that.

The only thing that matters is whether [`pcs.resource.resource_relocate_show`](https://github.com/ClusterLabs/pcs/blob/15ca9f88520b6d919a14ef6121c69643d2c244ce/pcs/resource.py#L3290) can parse it

clumens commented:
> The only thing that matters is whether [`pcs.resource.resource_relocate_show`](https://github.com/ClusterLabs/pcs/blob/15ca9f88520b6d919a14ef6121c69643d2c244ce/pcs/resource.py#L3290) can parse it

It looks like moving the resource management disabled message to under the "Current cluster status:" header will break things.  I'll think about how that can be fixed.

clumens commented:
Much closer, but still a few things left to do:

* Deal with return codes from pcmk__simulate.
* Obsess over the public API portions of this.
* Probably some more interactive testing (see the Test: commit for a couple ideas).
* Interactive testing around cib__signon_query error handling. 
* Verify we're not breaking pcs.

I guess I could probably make this no longer be a draft.  I'm happy with how a lot of this is working out.  I think it's going to be a big improvement.

kgaillot commented on lib/pacemaker/pcmk_simulate.c:
```diff
@@ -512,3 +512,27 @@ pcmk__simulate(pe_working_set_t *data_set, pcmk__output_t *out, pcmk_injections_
     return rc;
 }
+
+#ifdef BUILD_PUBLIC_LIBPACEMAKER
```
We don't need to guard this function. Only the stonith APIs are guarded b/c we don't want to expose stonith_t in the high-level API. (I was hoping we could get to replacing stonith_t with pcmk_ipc_api_t, but that's probably a long ways off -- it might be sufficient to make the stonith_t connection inside the high-level APIs rather than pass it to them.)

clumens replied:
Ah, for some reason I thought the define was hiding all these function until we'd had long enough to think about them and decide they were good enoguh to expose as public API.  I guess we'd better think more about whether these are correct, then.

kgaillot reviewed (commented):
The tests are a nice addition :)

If you're happy with some of the commits as-is, feel free to separate them into their own PR

For sufficiently descriptive return codes, I'm thinking of a few possibilities:
* Just return the errno and hope the user can figure out which it applies to
* The graph and dot files are secondary to the main purpose of the simulation, so we could lump any failures there into new codes e.g. pcmk_rc_graph_error and pcmk_rc_dot_error (which would be returned only if the rest of the function succeeded)
* Maybe pcmk__simulate() is too high-level, and we should separate it into stuff-before-graph-and-dot(), graph(), dot(), stuff-after-graph-and-dot()

kgaillot commented:
I forgot one thing: each public API addition should have an individual commit with message "API: <libraryname>: add <whatever>" for ease of creating the change log

clumens commented on lib/pacemaker/pcmk_sched_transition.c:
```diff
@@ -405,27 +402,27 @@ modify_configuration(pe_working_set_t * data_set, cib_t *cib,
 
         /* crm_xml_add(top, XML_ATTR_DC_UUID, dc_uuid);      */
-        crm_xml_add(top, XML_ATTR_HAVE_QUORUM, quorum);
+        crm_xml_add(top, XML_ATTR_HAVE_QUORUM, "1");
```
Ah, that's too bad.  I've removed this patch.

clumens commented:
I've added a patch to help with error codes in pcmk__simulate, though I haven't decided if it's sufficient or not.  I've also fixed an error about passing pcmk_sim_flags_e around and done some interactive testing to verify tools error out how they should.

kgaillot commented on lib/pacemaker/pcmk_sched_transition.c:
```diff
@@ -849,7 +846,7 @@ run_simulation(pe_working_set_t * data_set, cib_t *cib, GList *op_fail_list)
+        return -graph_rc;
     }
-    return 0;
+    return pcmk_rc_ok;
```
run_simulation() doesn't return a standard or legacy code, it returns enum transition_status on success (0 == transition_active) and -1 on error. So, we weren't exactly using it correctly in crm_simulate, but it worked more or less coincidentally.

It's internal, so we can change it ... we should return the enum explicitly rather than int, return graph_rc without modification, and have the caller compare the result to transition_complete

clumens replied:
Yeah, that's much better.  Fixed.

kgaillot commented on include/crm/common/results.h:
```diff
@@ -107,6 +107,9 @@ enum pcmk_rc_e {
     /* When adding new values, use consecutively lower numbers, update the array
      * in lib/common/results.c, and test with crm_error.
      */
+    pcmk_rc_sim_failed          = -1031,
```
Let's call this pcmk_rc_invalid_transition since that's how it will appear to users

kgaillot commented on lib/common/results.c:
```diff
@@ -250,7 +250,19 @@ static struct pcmk__rc_info {
+      -pcmk_err_generic,
+    },
+    { "pcmk_rc_sim_failed",
+      "Error occurred while running cluster simulation",
```
Maybe: Cluster simulation produced invalid transition

kgaillot commented on include/pacemaker.h:
```diff
@@ -26,6 +26,19 @@ extern "C" {
+    pcmk_sim_show_scores      = 1 << 3,
+    pcmk_sim_show_utilization = 1 << 4,
+    pcmk_sim_simulate         = 1 << 5,
+} pcmk_sim_flags_e;
```
If we want to standardize on typedef'ing enums, then the \_e name ending is a good idea; if we want to standardize on not typedef'ing, then we don't need \_e

clumens replied:
Fixed.

kgaillot reviewed (commented):
Almost there
